Allow ExternalTaskSensor to wait for taskgroup

xinbinhuang opened the pull request:
<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

closes: #14563

This PR enables ExternalTaskSensor to also wait for the external task_group. 

The implementation is to retrieve the external DAG from the DagBag and then check if the TaskGroup exists. If so, query and wait for the states of all tasks within that TaskGroup during the poking cycle.


---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

xinbinhuang committed:
Allow ExternalSensor to wait for taskgroup

xinbinhuang commented on airflow/sensors/external_task.py:
```diff
@@ -164,18 +179,23 @@ def poke(self, context, session=None):
             count_failed = self.get_count(dttm_filter, session, self.failed_states)
 
-        if count_failed == len(dttm_filter):
+        if 0 < count_failed <= len(dttm_filter):
```
Here I am making the assumption that as long as there is at least one external task failure, then we will want to fail the sensor. Though this changes the original behavior, I wonder in what situation users will allow partial failure?

xinbinhuang commented on airflow/sensors/external_task.py:
```diff
@@ -206,29 +228,48 @@ def get_count(self, dttm_filter, session, states) -> int:
+                .filter(TI.task_id.in_(external_task_group_task_ids))
                 .scalar()
-            )
+            ) / len(external_task_group_task_ids)
```
This is to keep the poking check behavior `return count_allowed == len(dttm_filter)`

xinbinhuang commented on airflow/sensors/external_task.py:
```diff
@@ -134,20 +146,23 @@ def __init__(
         self.external_dag_id = external_dag_id
+        self.external_task_group_id = external_task_group_id
         self.external_task_id = external_task_id
         self.check_existence = check_existence
```
`self.check_existence = check_existence` is `False` by default, which maybe make sense for external_dag or external_task. But external_task_group has to check and get an existing dag in order to get the list of task_ids. 

https://github.com/apache/airflow/blob/fce49402461ee4e7a5f6ffd18cee3121f3496a39/airflow/sensors/external_task.py#L174-L180

I wonder if we can change the default to True or even have `check_existence` enabled required? This can give more useful errors if the external task/dag does not exist as well as having a consistent behavior as external_task_group. Also, what would be use case to have a Sensor waiting for an object that doesn't exist until it times out?

xinbinhuang commented on airflow/sensors/external_task.py:
```diff
@@ -206,29 +232,48 @@ def get_count(self, dttm_filter, session, states) -> int:
+                f'DAG {self.external_dag_id} does not exist.'
+            )
+        task_ids = [task.task_id for task in task_group]
+        return task_ids
```
The main piece that you retrieve a list of tasks for a `TaskGroup`

kaxil commented on airflow/models/dag.py:
```diff
@@ -745,10 +745,18 @@ def tasks(self, val):
 
+    @property
+    def task_group_dict(self) -> Dict[str, "TaskGroup"]:
+        return {k: v for k, v in self._task_group.get_task_group_dict().items() if k is not None}
```
Since `.task_group` does the same

```suggestion
        return {k: v for k, v in self.task_group.get_task_group_dict().items() if k is not None}
```

xinbinhuang committed:
Use public method

Co-authored-by: Kaxil Naik <kaxilnaik@gmail.com>

xinbinhuang committed:
Remove unittest.TestCase

xinbinhuang committed:
Fix test: clear dag_runs for test_external_dag_sensor

xinbinhuang committed:
Test external task group sensor

fixup! Test external task group sensor

fixup! fixup! Test external task group sensor

fixup! fixup! fixup! Test external task group sensor

fixup! fixup! fixup! fixup! Test external task group sensor

xinbinhuang commented on airflow/sensors/external_task.py:
```diff
@@ -164,18 +184,23 @@ def poke(self, context, session=None):
             count_failed = self.get_count(dttm_filter, session, self.failed_states)
 
-        if count_failed == len(dttm_filter):
+        if count_failed > 0:
```
Here I am making the assumption that as long as there is at least one external task failure, then we will want to fail the sensor. Though this changes the original behavior, I think this will be a better.

turbaszek replied:
Should we add this comment in code?

xinbinhuang replied:
I don't think it's necessary more than an entry in the `UPDATING.md`. I think the only situation where you will have multiple counts is when the `execution_date_fn` returns more than one execution date to wait for. However, the original behavior will get you into a weird state when only part of the TIs fail, i.e. one fail and one succeeds, resulting in time out. IMHO, I think this's more like a bug than intended behavior. WDYT?

xinbinhuang commented on airflow/sensors/external_task.py:
```diff
@@ -206,29 +232,48 @@ def get_count(self, dttm_filter, session, states) -> int:
+
+        return query
+
+    def get_external_task_group_task_ids(self, session):
+        """Return task ids for the external TaskGroup"""
+        refreshed_dag_info = DagBag(read_dags_from_db=True).get_dag(self.external_dag_id, session)
+        task_group: Optional["TaskGroup"] = refreshed_dag_info.task_group_dict.get(
+            self.external_task_group_id
+        )
+        if not task_group:
+            raise AirflowException(
+                f"The external task group {self.external_task_group_id} in "
+                f"DAG {self.external_dag_id} does not exist."
+            )
+        task_ids = [task.task_id for task in task_group]
+        return task_ids
+
```
The main piece that you retrieve a list of tasks for a TaskGroup. I believe that `read_dags_from_db=True` is safe to use here because serialized dag is enabled by default in 2.0.

yuqian90 replied:
The existing task execution code is creating DagBag on its own instead of reading serialized dags from db. For example this line is creating a DagBag. I think we should do the same here. It's important for tasks to get the latest view of the dag during execution. 

https://github.com/apache/airflow/blob/f1edc220d3f9cb050016d23246a682276bd09eee/airflow/sensors/external_task.py#L213

turbaszek commented on airflow/models/dag.py:
```diff
@@ -1525,8 +1533,11 @@ def filter_task_group(group, parent_group):
 
         return dag
 
+    def has_task_group(self, group_id: str):
```
```suggestion
    def has_task_group(self, group_id: str) -> bool:
```

yuqian90 commented on airflow/models/dag.py:
```diff
@@ -745,10 +745,18 @@ def tasks(self, val):
 
+    @property
+    def task_group_dict(self) -> Dict[str, "TaskGroup"]:
+        return {k: v for k, v in self.task_group.get_task_group_dict().items() if k is not None}
```
`get_task_group_dict()` is a recursive function that can be costly. I think we should keep it a method instead of making it a property (which tends to suggest to users that it's cheap to access).

yuqian90 commented on airflow/models/dag.py:
```diff
@@ -745,10 +745,18 @@ def tasks(self, val):
 
+    @property
+    def task_groups(self) -> List["TaskGroup"]:
+        return list(self.task_group_dict.values())
```
Same here for `task_groups`

yuqian90 commented on airflow/models/dag.py:
```diff
@@ -1201,7 +1209,7 @@ def clear(
 
         if include_parentdag and self.is_subdag and self.parent_dag is not None:
-            p_dag = self.parent_dag.sub_dag(
+            p_dag = self.parent_dag.partial_subset(
```
Since you are changing this to `partial_subset`, let's use the new feature of this method. You may consider passing it like this to avoid the regex dance: 
```
task_ids_or_regex=[self.dag_id.split('.')[1]]
```

This is what the docstr of `partial_subset` says:
```
        :param task_ids_or_regex: Either a list of task_ids, or a regex to
            match against task ids (as a string, or compiled regex pattern).
```

yuqian90 commented on airflow/models/dag.py:
```diff
@@ -1282,7 +1290,7 @@ def clear(
                             if not external_dag:
                                 raise AirflowException(f"Could not find dag {tii.dag_id}")
-                            downstream = external_dag.sub_dag(
+                            downstream = external_dag.partial_subset(
```
Same here. This wants an exact match of the task_id. So passing `task_ids_or_regex` as a list of task_id is better.

yuqian90 commented on tests/sensors/test_external_task.py:
```diff
@@ -24,32 +23,49 @@
+    def setup_method(self):
         self.args = {'owner': 'airflow', 'start_date': DEFAULT_DATE}
         self.dag = DAG(TEST_DAG_ID, default_args=self.args)
+        SerializedDagModel.write_dag(self.dag)
```
I don't think `SerializedDagModel.write_dag` is needed if the task creates its own dagbag like it previously does.

yuqian90 commented on tests/sensors/test_external_task.py:
```diff
@@ -445,7 +545,7 @@ def clear_tasks(dag_bag, dag, task, start_date=DEFAULT_DATE, end_date=DEFAULT_DA
     Clear the task and its downstream tasks recursively for the dag in the given dagbag.
     """
-    subdag = dag.sub_dag(task_ids_or_regex=f"^{task.task_id}$", include_downstream=True)
+    subdag = dag.partial_subset(task_ids_or_regex=f"^{task.task_id}$", include_downstream=True)
```
same here. `task_ids_or_regex` can be a list of task_id

yuqian90 reviewed (changes requested):
Looks good overall. Some comments inline. Thanks for doing this.

xinbinhuang committed:
Update airflow/models/dag.py

Co-authored-by: Tomek Urbaszek <turbaszek@gmail.com>
